Dturner/pre-receive

novalis opened the pull request.

abliss commented on README.md:
```diff
@@ -159,10 +159,29 @@ more information, or one of our test drivers (initially at least
+  - directly referenced by the sub-repository's ref named
+    refs/heads/meta-$METASHA, or
+
+  - part of a chain of commits, the top commit of wich meets the
```
s/wich/which

abliss commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,227 @@
+    *urlToLocalPath(repo, url) {
+        const config = yield repo.config();
+        const subrepoUrlBase =
+            yield config.getStringBuf("gitmeta.subrepourlbase");
```
I strongly dislike the smashing together of words with no separator or case-change.

novalis replied:
My recollection from the git.git source code is that that style is what the APIs required (or at least what was always passed to them).  So I don't want to break anything.

abliss commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    getSyntheticBranchForCommit(commit) {
+        return "refs/git-meta/synthetic/meta-" + commit.id();
```
This doesn't seem to agree with README.md, which says `/refs/heads/meta-$METASHA`

abliss commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    *checkTree(repo, commit, parentTrees, tree) {
```
This function needs a comment.

abliss commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,227 @@
+            return true;
+        }
+
+        const ok = yield GitUtil.readNote(repo, NOTES_REF, commit.id());
```
Is it worth it to check that the note says "ok"? Could the note ever say something else? Should we even be writing "ok" to the note instead of an empty string?

abliss commented on lib/util/git_util.js:
```diff
@@ -249,6 +249,28 @@ exports.resolveCommitish = co.wrap(function *(repo, commitish) {
+    try {
+        return yield NodeGit.Note.read(repo, notesRef, oid);
+    }
+    catch (e) {
```
Doesn't `catch` go on the same line, like `else`?

novalis replied:
In this file, for some reason, no.

novalis commented:
I just pushed a few changes; they dramatically increase test coverage (and fix some of the bugs I discovered while doing so) .  Note that this code won't work with current node-git; it requires two patches to node-git ( https://github.com/nodegit/nodegit/pull/1080 and https://github.com/nodegit/nodegit/pull/1079 ) , only one of which is merged and neither of which is released.  I tried to make git-meta use the version of nodegit from git, but I couldn't figure it out.

abliss commented:
LGTM

bpeabody commented:
I think you need to add `split` to the list of dependencies. After doing that I get the following test failures:

```
  1) pre-receive simplest:
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.


  2) pre-receive notes block previous (even bad) history:
     UserError: fail
      at UserError (lib/util/user_error.js:47:26)
      at test/util/pre-receive.js:61:19
      at next (native)
      at onFulfilled (node_modules/co/index.js:65:19)

  3) pre-receive old shas block previous (even bad) history:
     UserError: fail
      at UserError (lib/util/user_error.js:47:26)
      at test/util/pre-receive.js:61:19
      at next (native)
      at onFulfilled (node_modules/co/index.js:65:19)

  4) pre-receive with a submodule, at meta commit:
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.


  5) pre-receive with a submodule in a subdir, at meta commit:
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.


  6) pre-receive with a submodule in a subdir, from earlier meta-commit:
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.


  7) pre-receive with a submodule in a subdir, irrelevant change:
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.
```

bpeabody commented:
A few random tabs in the files (I would expect `jshint` to catch these; let me see why it doesn't)

bpeabody commented:
I'm a little curious about this exception eating in the test driver:

```JavaScript
        try {
            const from = yield x.getReferenceCommit("from");
            old = from.id().toString();
        } catch (e) {
        }
```
Maybe document what's going on there?

bpeabody commented:
Otherwise, I don't see any obvious reason for the tests to time out. Sorry I have taken so long to look at this; will help out if you need it debugging the test.

novalis commented:
Are you testing against the patched nodegit?  

The timeout is not good -- there's a thing I don't understand about how promises, generators, etc work, and as a result I've written code which, instead of doing something sensible on exceptions, just times out.

bpeabody commented:
No, I am not testing against the patched version, let me try that.

novalis commented:
BTW, now that nodegit has merged my patches (but has not yet done a release), we could change the packages.json to point to some recent SHA of nodegit.  What do you think?

bpeabody commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,255 @@
+    }
+}
+
+PreReceiveCheck.prototype.checkUpdate = co.wrap(
```
There be tabs in this method.

bpeabody commented on lib/util/repo_ast_util.js:
```diff
@@ -652,6 +710,7 @@ exports.cloneRepo = function (original, url) {
             })
         },
         head: head,
+        notes: deepcopy(original.notes),
```
The constructor of `RepoAST` does a copy; you don't need to do it here too.

novalis replied:
it didn't but it does now.

bpeabody commented on test/util/shorthand_parser_util.js:
```diff
@@ -47,6 +47,7 @@ describe("ShorthandParserUtil", function () {
                 branches: {},
                 remotes: {},
                 index: {},
+                notes:{},
```
I know this is nit-picky, but a space between `:` and `{`, please.

novalis replied:
eslint has an option to enforce this; jshint apparently doesn't. So that's one argument for switching.

novalis commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,255 @@
+
+class PreReceiveCheck {
+
+    *urlToLocalPath(repo, url) {
```
I actually want everything else to be private -- what's the way to do that?

bpeabody replied:
Unfortunately, this is one of the suckier parts of JavaScript. You can get something close to private with ES6 symbols: http://stackoverflow.com/questions/22156326/private-properties-in-javascript-es6-classes, or you can do it the "old school" way with closures as I did with the `SubmoduleRebaser` in `lib/util/rebase.js`.

novalis commented on lib/util/read_repo_ast_util.js:
```diff
@@ -36,6 +36,7 @@
  */
 
 const assert   = require("chai").assert;
+const async     = require("generator-async");
```
co converts a function that returns a generator into a function that returns a promise. I need to convert a function that returns a generator into a sync function.  But I guess I can do it another way.

bpeabody replied:
I think `async.run(foo)` is the equivalent of `co(foo)`. If `async` provides something that `co` doesn't, feel free to keep using `async`.

novalis replied:
oh, right.  The problems is not that co() doesn't work -- it's that errors get swallowed either way, and there were some bugs.  Yay.

novalis commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,255 @@
+        const localPath = path.join(subrepoRootPath, remotePath);
+        if (localPath[0] === "/") {
+            return localPath;
+        } else {
```
sorry, does this entail a change here?

novalis commented:
actually I should fix this to match the refs/commits/$commit instead of whatevs/$metacommit.  So don't merge quite yet.

novalis commented:
hope that doesn't make testing super-unfun.

novalis commented:
... actually, the tests mocked that.

novalis commented:
Ok, let me know how this one looks

bpeabody commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,315 @@
+         * commits that correspond to this tree
+         * @param {NodeGit.Tree} tree The tree to check
+         */
+        this[checkTree] = function *(repo, commit, parentTrees, tree) {
```
I believe libgit2 provides a way to look at what's changed in a commit, e.g. `NodeGit.Commit.getDiffWithOptions` that might be simpler (you won't need this recursive tree processing logic) and more efficient. Also, since we really only care about submodules you could probably use `SubmoduleUtil.getSubmoduleChanges`, which will tell you want submodules have been added/changed/removed in a commit. It looks like its documentation is incorrect in that it returns sets, not maps; I'll change.

novalis commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,315 @@
+function flatmap(arr, fn) {
+    return arr.map(fn).filter(identity);
+}
+
```
Maybe, but ramda's "chain" isn't quite the same as flatmap (or maybe I'm abusing the flatmap terminology); it treats a null return as a one-element list instead.

novalis commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,315 @@
+const checkSubmodule = Symbol();
+const checkTree = Symbol();
+const parentLoop = Symbol();
+class PreReceiveCheck {
```
If I don't do that, then how do I mock methods?

bpeabody replied:
What methods do you need to mock and for what purpose? I don't see us providing a mock implementation of `PreReceive`. Actually, it might make sense to name this util something about its purpose, like `validate_meta_util` rather than its current caller.

novalis replied:
getSyntheticBranchForCommit for testing.  Although actually I might not need to do that now that I no longer care about the meta commit id, so maybe I can simplify htat.

novalis replied:
Oh wait we still don't know the actual shas of the subrepos, so we can't create the synthetic metabranches, so we do need the class.  But will rename.

bpeabody replied:
Still not sure what you need the class for (not that I have an extremely string opinion on the subject). I think you would need one only if you have state.

novalis commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,315 @@
+                return true;
+            }
+            handled[commit.id()] = 1;
+            if (oldSha === commit.id().toString()) {
```
new refs have an oldSha of 0000000000000000000000000000000000000000

novalis commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,315 @@
+
+                function getEntryByEntryName(parentTree) {
+                    try {
+                        return parentTree.entryByName(entry.name());
```
I can add an assert for that case, but I bet it fails anyway.

bpeabody replied:
Let me rephrase:
1. What case have you encountered that necessitated this try/catch.
2. If it was from a malformed tree (I remember our magic mirror used to write invalid trees), should we explicitly reject it in the same way we would if missing a synthetic meta-branch?

novalis replied:
I think this is a normal case; a parent tree might just not have this entry (maybe it's newly added).

novalis commented:
OK, that submodule diff thingy simplified the code. I wonder why I wrote it the complicated way in the first place.

And I rearranged the commits to be a bit cleaner.  I think I addressed all comments.

Oh, I renamed "synthetic metabranch" to "synthetic branch", since they're now not as related to the meta repo.

novalis commented on lib/util/pre_receive_util.js:
```diff
@@ -0,0 +1,315 @@
+        const newCommit = yield repo.getCommit(newAnnotated.id());
+        const success = yield this[parentLoop](repo, newCommit, oldSha, {});
+        if (success) {
+            yield NodeGit.Note.create(repo, NOTES_REF, newCommit.committer(),
```
I think we'll rarely end up replicating too much work.

bpeabody commented:
I'll take a fresh look at everything.
